amivapi/auth/apikeys.py

```diff
@@ -3,38 +3,107 @@
 
 from flask import abort, current_app, g
 
-from ruamel import yaml
+from amivapi.utils import register_domain
+
+from .auth import AmivTokenAuth
 
 
 def authorize_apikeys(resource):
-    """Check if user is an apikey, and if it is, do authorization."""
-    permissions = [
-        data['permissions'] for data in current_app.config['APIKEYS'].values()
-        if data['token'] == g.get('current_token')]
-    if permissions:
-        # We can just take first (and only) element since token is unique
-        permission = permissions[0].get(resource)
+    """Check if user is an apikey, and if it is, do authorization.
+
+    Also update 'updated' timestamp everytime a key is accessed
+    """
+    apikeys = current_app.data.driver.db['apikeys']
+    apikey = apikeys.find_one({'token': g.get('current_token')},
+                              {'permissions': 1})
+
+    if apikey:
+        # Get permission for resource if they exist
+        permission = apikey['permissions'].get(resource)
+
+        # Update timestamp (remove microseconds to match mongo precision)
+        new_time = dt.utcnow().replace(microsecond=0)
+        apikeys.update_one({'_id': apikey['_id']},
+                           {'$set': {'_updated': new_time}})
 
         if permission == 'read':
             g.resource_admin_readonly = True
         elif permission == 'readwrite':
             g.resource_admin = True
         elif g.get('auth_required'):
-            abort(403, "The APIKEY does not grant the required permissions.")
+            abort(403, "The API key exists but does not grant the required "
+                       "permissions.")
+
+
+class ApikeyAuth(AmivTokenAuth):
+    """Do not let (non-admin) users view API keys at all.
+
+    We don't want to filter the output , we want to abort completely."""
+    def create_user_lookup_filter(self, user):
+        abort(403)
+
+
+apikeydomain = {
+    'apikeys': {
+        'description': "API Keys can be used to given permissions to other "
+                       "applications.",
+
+        'public_methods': [],
+        'public_item_methods': [],
+        'resource_methods': ['GET', 'POST'],
+        'item_methods': ['GET', 'PATCH', 'DELETE'],
+
+        'authentication': ApikeyAuth,
+
+        'schema': {
+            'name': {
+                'type': 'string',
+                'required': True,
+                'nullable': False,
+                'empty': False,
+                'description': 'A name to identify the key.',
+                'unique': True
+                },
+            'token': {
+                'type': 'string',
+                'readonly': True,
+                'description': 'The token that can be used for auth. '
+                'Will be randomly generated for every new key.'
+            },
+            'permissions': {
+                'type': 'dict',
+                'propertyschema': {'type': 'string',
+                                   'api_resources': True},
```

Comment:
This validator is removed from groups/validation.py further down. How does this even work?